WordPress/src/main/java/org/wordpress/android/ui/posts/PostListMainViewModel.kt

```diff
@@ -90,4 +134,89 @@ class PostListMainViewModel @Inject constructor(
             }
         }
     }
+
+    private fun getAuthorFilterItems(selection: AuthorFilterSelection): List<AuthorFilterListItemUIState> {
+        return AuthorFilterSelection.values().map { value ->
+            @ColorRes val backgroundColorRes: Int =
+                    if (selection == value) R.color.grey_lighten_30_translucent_50
+                    else R.color.transparent
+
+            when (value) {
+                ME -> AuthorFilterListItemUIState.Me(accountStore.account?.avatarUrl, backgroundColorRes)
+                EVERYONE -> AuthorFilterListItemUIState.Everyone(backgroundColorRes)
+            }
+        }
+    }
+
+    /**
+     * Only the non-null variables will be changed in the current state
+     */
+    private fun updateViewState(
+        isFabVisible: Boolean? = null,
+        isAuthorFilterVisible: Boolean? = null,
+        authorFilterSelection: AuthorFilterSelection? = null,
+        authorFilterItems: List<AuthorFilterListItemUIState>? = null
+    ) {
+        IllegalArgumentException().printStackTrace()
+        val currentState = viewState.value
+
+        if (currentState == null) {
+            if (BuildConfig.DEBUG) {
+                throw IllegalStateException("updateViewState should not be called before the initial state is set")
+            }
+
+            return
+        }
+
+        _viewState.value = PostListMainViewState.Idle(
+                isFabVisible ?: currentState.isFabVisible,
+                isAuthorFilterVisible ?: currentState.isAuthorFilterVisible,
+                authorFilterSelection ?: currentState.authorFilterSelection,
+                authorFilterItems ?: currentState.authorFilterItems
+        )
+
+        if (authorFilterSelection != null && currentState.authorFilterSelection != authorFilterSelection) {
+            _updatePostsPager.value = authorFilterSelection
+        }
+    }
+
+    sealed class PostListMainViewState(
```

Comment:
This sealed class doesn't seem to add any value over a regular class. Sealed classes in Kotlin are Algebraic Data Types, meaning an instance can be one of a finite number of options. You can think of them as glorified enums. This gives us a lot of confidence in our types because if a new type is added to a sealed class, the compiler will warn us if we use `when` properly. (meaning we don't do a catch all with _else_) I'd strongly suggest reading a bit more about sealed classes before proceeding as that'll be valuable in other parts of this PR as well.

For this class specifically, I suggest we use a simple data class since we are dealing with a single type.

Reply:
I am aware about what sealed classes are and should be used for.

I agree it doesn't add value here, however I used to for the exact reason you mentioned. 
> This gives us a lot of confidence in our types because if a new type is added to a sealed class, the compiler will warn us if we use when properly.
It will make adding a new ui state types easy and give us warning is `PostListActivity` when that base state is changed.

I can change it to a simple class if you would really prefer. Please confirm.

Reply:
> however I used to for the exact reason you mentioned.

Hmm.. maybe I am missing something here but as far as I can tell we are not currently using these classes with `when` or even with an `if`, so if you don't mind, could you please elaborate on how it's current usage is related to the reason I have mentioned?

Reply:
It was more the simplicity of adding a new type, where a developer might move some of the current variables to the child state I was thinking about.
My thought was that it clearly shows a future developer that we would expect different states to be a child of this sealed class. It is simpler cognitively to think "I add my new state here" than "I'll convert this to a sealed class move the add two states, one for the current state and one for my new one"

Just my opinion though.

Reply:
I think that's a pretty slippery slope because in that case we'd make everything a `sealed class`. My take is if a code has no current value, it should not be there. If we end up having another state, then we'll convert it to the `sealed class`.

Please excuse me if I am misunderstanding your arguments, but I feel like this contradicts your comment [here](https://github.com/wordpress-mobile/WordPress-Android/pull/9484#discussion_r269361837) a little because in that case we clearly have 2 different states and proper inheritance which is exactly what a `sealed class` address. On the other hand we neither have inheritance nor multiple states.

Assuming I am not misunderstanding anything, I'd strongly suggest converting this to a simple `data class` which will allow us to `copy` values which I think might come handy when we try to address the update view state issue.

Reply:
The comment you linked to I don't believe does contradict what I'm saying. Since the view holders used in `AuthorSelectionAdapter` are never used in conjunction with each other.

One is used for the spinner's view when it is displayed, and the other is used in the dropdown. There is no conceivable reason in the future we would use the base `ViewHolder` class proposed in a when or if.


In regards to the code this thread is referring to, I'll make the change requested.

Reply:
> The comment you linked to I don't believe does contradict what I'm saying. Since the view holders used in AuthorSelectionAdapter are never used in conjunction with each other.

> One is used for the spinner's view when it is displayed, and the other is used in the dropdown. There is no conceivable reason in the future we would use the base ViewHolder class proposed in a when or if.

That makes sense, I stand corrected. Thank you for the explanation.

Reply:
Resolved in 4504258bfa45a1266535111922280c628de809ba

Reply:
I would have probably gone with a data class as Oguz suggested, but I think it doesn’t really matter in this case.